### PR TITLE
nan detection and intervention 

### LINKED ISSUE #1008: make metric-comparison in ModelCheckpoint robust to NaN
## 🐛 Bug

When the metric that is used in `ModelCheckpoint` reaches `NaN` in one epoch and then returns to a number in the following epoch, the model will not be saved as comparisons to `NaN` always return `False`.
 
Here a screenshot from my training:

Screenshot from 2020-03-02 14-31-46

### Expected behavior
I think it would make sense to no save the model if the metric is `NaN`.

### Environment

 - PyTorch Version (e.g., 1.0): 1.2.0
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.7
 - CUDA/cuDNN version: 10.1

### PR DESCRIPTION
**Pull request by awaelchli**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #1008.

- Detection of non-finite loss or parameter values
- The check for nan or inf is done before optimization
- When such values are detected, the training stops immediately.

**Open question:** what happens to the `print_nan_grads` Trainer argument? is it still useful?
- For now, I removed the `print_nan_grads` argument and print them whenever nan is detected. Since training stops anyway, it wouldn't make sense to have this arg.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
hey there, we have added GPU CI test, so could we kindly ask to rebase/merge master which will trigger these tests so we do not need to test it manually... Thx for your understanding :robot:

**Comment by awaelchli** [comment]
I think this is close to finished, just need review and core developers to comment/decide about the open question in the description above. Thanks.

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -208,7 +208,10 @@ def __init__(
 
             precision: Full precision (32), half precision (16).
 
-            print_nan_grads: Prints gradients with nan values
+            print_nan_grads:
+                .. warning:: .. deprecated:: 0.7.0
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]:
```suggestion
                .. warning:: .. deprecated:: 0.7.2
```

File: pytorch_lightning/trainer/training_tricks.py [hunk 2]
```diff
@@ -45,12 +48,31 @@ def clip_gradients(self):
...
+        if not torch.isfinite(loss).all():
+            sys.exit(
+                'The loss returned in `training_step` is nan or inf.'
+                ' Will stop training.'
+            )
+        # check if a network weight is nan
+        for name, param in model.named_parameters():
+            if not torch.isfinite(param).all():
+                self.print_nan_gradients()
+                sys.exit(
```
Comment (Borda) [inline pytorch_lightning/trainer/training_tricks.py]:
```suggestion
                raise ValueError(
```
or any other suitable Exception...

File: pytorch_lightning/trainer/training_tricks.py [hunk 3]
```diff
@@ -45,12 +48,31 @@ def clip_gradients(self):
...
         for param in model.parameters():
             if (param.grad is not None) and torch.isnan(param.grad.float()).any():
                 log.info(param, param.grad)
 
+    def detect_nan_tensors(self, loss: Tensor) -> None:
+        model = self.get_model()
+
+        # check if loss is nan
+        if not torch.isfinite(loss).all():
+            sys.exit(
```
Comment (Borda) [inline pytorch_lightning/trainer/training_tricks.py]:
```suggestion
            raise ValueError(
```
or any other suitable Exception...

File: tests/test_cpu_models.py [hunk 4]
```diff
@@ -359,5 +357,63 @@ def test_single_gpu_model(tmpdir):
...
+def test_nan_loss_detection(tmpdir):
+    test_step = 8
+
+    class InfLossModel(LightTrainDataloader, TestModelBase):
+
+        def training_step(self, batch, batch_idx):
+            output = super().training_step(batch, batch_idx)
+            if batch_idx == test_step:
+                if isinstance(output, dict):
+                    output['loss'] /= 0  # make loss infinite
```
Comment (Borda) [inline tests/test_cpu_models.py]: here may cry about division by 0, so rather set it as Inf directly

File: pytorch_lightning/trainer/training_tricks.py [hunk 5]
```diff
@@ -45,12 +48,31 @@ def clip_gradients(self):
...
         model = self.get_model()
         for param in model.parameters():
             if (param.grad is not None) and torch.isnan(param.grad.float()).any():
                 log.info(param, param.grad)
 
+    def detect_nan_tensors(self, loss: Tensor) -> None:
+        model = self.get_model()
+
+        # check if loss is nan
+        if not torch.isfinite(loss).all():
```
Comment (Borda) [inline pytorch_lightning/trainer/training_tricks.py]: so some can be Inf bot not all, right?

File: pytorch_lightning/trainer/training_tricks.py [hunk 6]
```diff
@@ -45,12 +48,31 @@ def clip_gradients(self):
...
+                ' Will stop training.'
+            )
+        # check if a network weight is nan
+        for name, param in model.named_parameters():
+            if not torch.isfinite(param).all():
+                self.print_nan_gradients()
+                sys.exit(
+                    f'Detected nan and/or inf values in `{name}`.'
+                    ' Check your forward pass for numerically unstable operations.'
+                    ' Will stop training.',
```
Comment (Borda) [inline pytorch_lightning/trainer/training_tricks.py]:
```suggestion
```

File: pytorch_lightning/trainer/training_tricks.py [hunk 7]
```diff
@@ -45,12 +48,31 @@ def clip_gradients(self):
...
                 log.info(param, param.grad)
 
+    def detect_nan_tensors(self, loss: Tensor) -> None:
+        model = self.get_model()
+
+        # check if loss is nan
+        if not torch.isfinite(loss).all():
+            sys.exit(
+                'The loss returned in `training_step` is nan or inf.'
+                ' Will stop training.'
```
Comment (Borda) [inline pytorch_lightning/trainer/training_tricks.py]:
```suggestion
```

File: pytorch_lightning/trainer/training_tricks.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_tricks.py]: the if condition is true when at least one element in the loss is either Nan, +inf, or -inf. Most of the time the loss is just one element (0-dim). But just in case that changes, I have added the .all()

File: tests/test_cpu_models.py (hunk 4 above)
Comment (awaelchli) [reply tests/test_cpu_models.py]: ok, thanks!

File: pytorch_lightning/trainer/training_tricks.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_tricks.py]: my bad overlooked `isfinate` as `infinite` lol
